### ISSUE TITLE
validator on prop change

### ISSUE THREAD

**Issue by hirokiosame** [vuejs/vue#5871]
### What problem does this feature solve?
The `validator` on props are executed against the passed-in props on creation, but if it was binding, the value of the prop can be updated to bypass validation.

### What does the proposed API look like?
No API change. Just validation checks on value change.

**Comment by yyx990803** [comment]
Validators are executed on prop change. If not, please provide a repro.

**Comment by hirokiosame** [comment]
Ah, my mistake. I encountered the case of passing in an array as a prop, and manipulating an element via `arr[0] = 'str';` where the validator of the prop enforces that all elements are numbers. I learned that the issue is with reactivity and that it must be updated via `this.$set(arr, '0', 'str')` for the component to acknowledge the change.

If object properties that are declared on data-creation can be reactive, why can't array indices be?
